ml-agents/mlagents/trainers/reward_providers/gail_reward_provider.py

```diff
@@ -0,0 +1,226 @@
+from typing import Optional
+import numpy as np
+import torch
+
+from mlagents.trainers.buffer import AgentBuffer
+from mlagents.trainers.reward_providers.base_reward_provider import BaseRewardProvider
+from mlagents.trainers.settings import GAILSettings
+from mlagents_envs.base_env import BehaviorSpec
+from mlagents.trainers.reward_providers.curiosity_reward_provider import (
+    actions_to_onehot,
+    Swish,
+)
+from mlagents.trainers.demo_loader import demo_to_buffer
+
+
+class GAILRewardProvider(BaseRewardProvider):
+    def __init__(self, specs: BehaviorSpec, settings: GAILSettings) -> None:
+        super().__init__(specs, settings)
+        self._discriminator_network = DiscriminatorNetwork(specs, settings)
+        _, self._demo_buffer = demo_to_buffer(
+            settings.demo_path, 1, specs
+        )  # This is supposed to be the sequence length but we do not have access here
+        params = list(self._discriminator_network.parameters())
+        self.optimizer = torch.optim.Adam(params, lr=settings.learning_rate)
+
+    def evaluate(self, mini_batch: AgentBuffer) -> np.ndarray:
+        with torch.no_grad():
+            estimates, _ = self._discriminator_network.compute_estimate(
+                mini_batch, use_vail_noise=False
+            )
+            return -torch.log(
+                1.0 - estimates * (1.0 - self._discriminator_network.EPSILON)
+            )
+
+    def update(self, mini_batch: AgentBuffer) -> None:
+        expert_batch = self._demo_buffer.sample_mini_batch(
+            mini_batch.num_experiences, 1
+        )
+        loss = self._discriminator_network.compute_loss(mini_batch, expert_batch)
+        self.optimizer.zero_grad()
+        loss.backward()
+        self.optimizer.step()
+
+
+class DiscriminatorNetwork(torch.nn.Module):
+    gradient_penalty_weight = 10.0
+    z_size = 128
+    alpha = 0.0005
+    mutual_information = 0.5
+    EPSILON = 1e-7
+    initial_beta = 1.0
+
+    def __init__(self, specs: BehaviorSpec, settings: GAILSettings) -> None:
+        super().__init__()
+        self._policy_specs = specs
+        self._settings = settings
+
+        # create the encoders for observations, actions and done
+        vec_obs_size = sum(
+            shape[0] for shape in specs.observation_shapes if len(shape) == 1
+        )
+        # vis_obs_shapes = [
+        #     shape for shape in specs.observation_shapes if len(shape) == 3
+        # ]
+
+        encoder_input_size = vec_obs_size  # + visuals later and normalization
+        if settings.use_actions and specs.is_action_continuous():
+            encoder_input_size += specs.action_size + 1  # + 1 is for done
+        if settings.use_actions and specs.is_action_discrete():
+            encoder_input_size += (
+                sum(specs.discrete_action_branches) + 1
+            )  # + 1 is for done
+
+        self.encoder = torch.nn.ModuleList(
```

Comment:
You can use a `torch.nn.Sequential` here and avoid the for loop later on in the code.